Complete overhaul of transform manual page (#12034)
rerun-io/rerun: docs/content/concepts/transforms.md

Note that by default, all entities are connected via identity transforms.

snippet: concepts/transform3d_hierarchy_simple

In this hierarchy:
- The `sun` entity exists at the origin of its own coordinate system
- The `sun/planet` transform places the planet 6 units from the sun, along the x-axis
- The `sun/planet/moon` transform places the moon 3 units along x away from the planet

This creates a transform hierarchy where transforms propagate down the entity tree. The moon's final position in the sun's coordinate system is 9 units away (6 + 3),
because the transforms are applied sequentially.

### Named transform frames

While entity path hierarchies work well for many cases, sometimes you need more flexibility in organizing your transforms.
In particular for anyone familiar with ROS we recommend using named transform frames as it allows you to model
your data much closer to how it would be defined when using ROS' [tf2](https://wiki.ros.org/tf2) library.

By explicitly specifying transform frames, you can decouple spatial relationships from the entity hierarchy.

> Instead of relying on entity path relationships, each entity is first associated with a named transform frame using

Reviewer: Ok, scratch that. I see now that you use "transform frame" thoughout the whole document. Let's favor consistency 😄

the [`CoordinateFrame`](https://rerun.io/docs/reference/types/archetypes/coordinate_frame) archetype.

The relationship between transform frames is then determined by logging [`Transform3D`](https://rerun.io/docs/reference/types/archetypes/transform3d)
with `child_frame` and `parent_frame` parameters to define the geometric relationship between two transform frames.

snippet: concepts/transform3d_hierarchy_named_frames

Note that unlike in ROS, you can log your transform relationship on _any_ entity.

> However, currently once a relation between two frames has been logged to an entity, this relation may no longer be logged on any other entity.

Reviewer: Would highlight / explain this a bit different, e.g.:
```suggestion
**Note:** A current limitation to this is that once a `Transform3D` relating two frames has been logged to an entity, this particular relation may no longer be logged on any other entity.
```

</picture>

## Other transform types

### Pinhole projections

In Rerun, pinhole cameras are not merely another archetype that can be visualized,
they are also treated as spatial relationships that define projections from 3D spaces to 2D subspaces.
This unified approach allows the Viewer to handle both traditional 3D-to-3D transforms and 3D-to-2D projections.

The [`Pinhole`](https://rerun.io/docs/reference/types/archetypes/pinhole) archetype defines this projection relationship through its intrinsic matrix (`image_from_camera`) and resolution.
Both implicit & named coordinate frames are supported, exactly as on [`Transform3D`](https://rerun.io/docs/reference/types/archetypes/transform3d).

With the right setup, pinholes allow a bunch of powerful visualizations:
* the pinhole glyph itself in 3D views
* 2D in 3D: all 2D content that is part of the pinhole's transform subtree
* 3D in 2D: if the pinhole is at the origin of the view, 3D objects can be projected through pinhole camera into the view.
    * Both the [nuscenes](https://rerun.io/examples/robotics/nuscenes_dataset) and [arkit](https://rerun.io/examples/spatial-computing/arkit_scenes) examples make use of this

> If a transform frame relationship has both a pinhole projection & regular transforms (in this context often regarded as the camera extrinsics)

Reviewer: missing comma
```suggestion
If a transform frame relationship has both a pinhole projection & regular transforms (in this context often regarded as the camera extrinsics),
```

---
title: Transforms & Transform Frames
order: 300
---

<!-- Figma file for diagrams in this article: https://www.figma.com/board/PTwJKgi9kQOqG7ZgzdhrDL/Transforms-doc-page-graphs?t=fWkOGxxn6mZkkCON-1 -->

> 

Reviewer 2: Note on overall layout. Right now it feels a little weird so maybe just rearranging the header hierarchy will help?

Maybe something like (no opinion on exact names):
* Transforms
   * Entity path transforms
   * Named transforms
   * Transforms - under the hood
* Special Transform Archetypes
   * Pinhole
   * View coordinates
   * Pose transforms

Rerun comes with built-in support for modeling spatial relationships between entities.
This page details how the [different archetypes](https://rerun.io/docs/reference/types/archetypes#transforms) involved interact with each other and explains how geometric transforms are set up in Rerun.

## Transform hierarchies with entity paths

The [`Transform3D`](https://rerun.io/docs/reference/types/archetypes/transform3d) archetype allows you to specify how one coordinate system relates to another through translation, rotation, and scaling.

> The simplest way to use transforms is through entity path hierarchies, where each transform describes the relationship between an entity and its parent path.

Reviewer 2: NIT: Link to entity path details?

TODO: make tested cross language snippet
```python
import rerun as rr
import numpy as np

rr.init("named_frames_example", spawn=True)
rr.log("/", rr.ViewCoordinates.RIGHT_HAND_Z_UP, static=True)

# Define entities with named coordinate frames
rr.log("sun", rr.Ellipsoids3D(centers=[0, 0, 0], half_sizes=[1, 1, 1], colors=[255, 200, 10]),
       rr.CoordinateFrame("sun_frame"))
rr.log("planet", rr.Ellipsoids3D(centers=[0, 0, 0], half_sizes=[0.4, 0.4, 0.4], colors=[40, 80, 200]),
       rr.CoordinateFrame("planet_frame"))
rr.log("moon", rr.Ellipsoids3D(centers=[0, 0, 0], half_sizes=[0.15, 0.15, 0.15], colors=[180, 180, 180]),
       rr.CoordinateFrame("moon_frame"))

# Connect the viewer to the sun's coordinate frame
rr.log("/", rr.CoordinateFrame("sun_frame"))

# Define frame relationships
rr.log("planet_transform", rr.Transform3D(
    translation=[6.0, 0.0, 0.0],
    child_frame="planet_frame",
    parent_frame="sun_frame"
))
rr.log("moon_transform", rr.Transform3D(
    translation=[3.0, 0.0, 0.0],
    child_frame="moon_frame",
    parent_frame="planet_frame"
))
```

Note that unlike in ROS, you can log your transform relationship on _any_ entity.

> However, currently once an entity specified the relation between two frames, this relation may no longer be logged on any other entity.

Reviewer 2: The flip flopping order made me have to reread

```suggestion
However, currently once a relation between two frames has been logged to an entity, this relation may no longer be logged on any other entity.
```

    rr.CoordinateFrame("tf#/robot"),
    rr.Transform3D(
        translation=[1, 0, 0],
        child_frame="tf#/robot",
        parent_frame="tf#/"
    )
)
rr.log("robot/arm",
    rr.CoordinateFrame("tf#/robot/arm"),
    rr.Transform3D(
        translation=[0, 1, 0],
        child_frame="tf#/robot/arm",
        parent_frame="tf#/robot"
    )
)
rr.log("robot/arm/gripper",
    rr.CoordinateFrame("tf#/robot/arm/gripper"),
    rr.Points3D([0, 0, 0])
)
```

<picture>
  <img src="https://static.rerun.io/transform_graph_translated/869b741ecce84c6b9af183922d32226a32a500bc/full.png" alt="">
  <source media="(max-width: 480px)" srcset="https://static.rerun.io/transform_graph_translated/869b741ecce84c6b9af183922d32226a32a500bc/480w.png">
  <source media="(max-width: 768px)" srcset="https://static.rerun.io/transform_graph_translated/869b741ecce84c6b9af183922d32226a32a500bc/768w.png">
  <source media="(max-width: 1024px)" srcset="https://static.rerun.io/transform_graph_translated/869b741ecce84c6b9af183922d32226a32a500bc/1024w.png">
</picture>

### Mixing named and implicit transform frames

We generally do not recommend mixing named and implicit transform frames since it can get confusing,

> but doing so works seamlessly and can be useful in some situations.

Reviewer 2: ```suggestion
but doing so works seamlessly and can be useful if necessary.
```

  <source media="(max-width: 768px)" srcset="https://static.rerun.io/pinhole-projections/ceb1b4124e111b5d0a786dd48909a1cbb52eca4c/768w.png">
  <source media="(max-width: 1024px)" srcset="https://static.rerun.io/pinhole-projections/ceb1b4124e111b5d0a786dd48909a1cbb52eca4c/1024w.png">
  <source media="(max-width: 1200px)" srcset="https://static.rerun.io/pinhole-projections/ceb1b4124e111b5d0a786dd48909a1cbb52eca4c/1200w.png">
  <img src="https://static.rerun.io/pinhole-projections/ceb1b4124e111b5d0a786dd48909a1cbb52eca4c/full.png">
</picture>

## View coordinates

You can use the [`ViewCoordinates`](https://rerun.io/docs/reference/types/archetypes/view_coordinates) archetype to set your preferred view coordinate systems, giving semantic meaning to the XYZ axes of the space.

For 3D spaces it can be used to log what the up-axis is in your coordinate system. This will help Rerun set a good default view of your 3D scene, as well as make the virtual eye interactions more natural. In Python this can be done with `rr.log("/", rr.ViewCoordinates.RIGHT_HAND_Z_UP, static=True)`.

> Note that in this example the archetype is logged at the root path, this will make it apply to all 3D views. Generally, a 3D view picks up view coordinates at or above its origin entity path.

Reviewer 2: How does this relate to logging the coordinate frame to root?

```
# Connect the viewer to the sun's coordinate frame
rr.log("/", rr.CoordinateFrame("sun_frame"))
```
